### ISSUE TITLE
can-stache 4.0

### ISSUE THREAD

**Issue by phillipskevin** [canjs/can-stache#289]
## TLDR

To make `stache` easier to lean and understand, we should:

- Remove implicit scope walking
- Remove automatic function calling
- Remove [`%special`](https://canjs.com/doc/can-stache/keys/special.html) keywords
- Remove `@`
- Use `can-parse` for parsing Expressions

> This was discussed at a [contributors meeting (28:44)](https://youtu.be/95OX6-qii24?t=28m44s)

## Steps

3.0 Issues:

- [x] [Deprecate `{{#each x as KEY}}` in favor of {{#each x item=value num=index}}](https://github.com/canjs/can-stache/issues/300)
- [x] [Warn when using implicit scope walking](https://github.com/canjs/can-stache/issues/311)
- [x] [Warn when using automatic function calling](https://github.com/canjs/can-stache/issues/312)
- [x] [Replace *variable and %special with `scope` (1st half of issue)](https://github.com/canjs/can-stache/issues/304)
- [x] [Update docs, guides, example apps to use Call Expressions for built-in helpers](https://github.com/canjs/canjs/issues/3589)
- [x] [Add `scope.root`](https://github.com/canjs/can-stache/issues/349)
- [x] [#each() doesn't work correctly with computes that wrap lists](https://github.com/canjs/can-stache/issues/365)
- [ ] [Implement scope.partials](https://github.com/canjs/can-view-scope/issues/122)

4.0 issues:
- [x] [Remove Legacy Binding Syntaxes](https://github.com/canjs/can-stache-bindings/issues/339)
- [x] [Remove `as` in `each`](https://github.com/canjs/can-stache/issues/363)
- [x] [Remove `registerSimpleHelper`](https://github.com/canjs/can-stache/issues/368)
- [x] [Remove `%special` keywords and `*reference` variables](https://github.com/canjs/can-stache/issues/313)
- [x] [Remove implicit scope walking](https://github.com/canjs/can-stache/issues/314)
- [x] [Allow people to opt-in to scope walking](https://github.com/canjs/can-view-scope/issues/116)
- [x] [Remove automatic function calling](https://github.com/canjs/can-stache/issues/315)
- [x] [Remove `@`](https://github.com/canjs/can-stache/issues/47)
- [x] [Remove magicMatch documentation](https://github.com/canjs/can-view-parser/issues/78)
- [x] [can we remove `options` scope](https://github.com/canjs/can-stache/issues/289#issuecomment-332964178)
- [x] [can we remove `looksLikeAhelper` weird logic](https://github.com/canjs/can-stache/issues/289#issuecomment-340062580)
- [x] [Make `helpers`, `partials`, `templates`, `optionsScope` available on templateContext  (2nd half of issue)](https://github.com/canjs/can-stache/issues/304)
- [x] [Remove remaining automatic function calling](https://github.com/canjs/can-stache/issues/401)

5.0 issues:
- [ ] [look into where we can improve performance](https://github.com/canjs/can-stache/issues/244)
- [ ] [Use `can-parse` for Expressions](https://github.com/canjs/can-stache/issues/290)
- [ ] [Don't encode attributes that are not part of bindings](https://github.com/canjs/can-view-parser/issues/77)
- [ ] [Allow faster lookup path that does not use `undefinedObserves`](https://github.com/canjs/can-view-scope/issues/117)

Possible other steps:

- [ ] [don't warn on properties that are defined on DefineMap, but are `undefined`](https://github.com/justinbmeyer)
- [ ] [refactor `can-stache-key` to better use `can-reflect` (might also cover issue above)](https://github.com/canjs/can-stache/issues/289#issuecomment-332965993)
- [ ] [should we only allow using Helper Expressions for Helpers (not for functions on the scope)](https://github.com/justinbmeyer)
- [ ] [[bindings] namespace internal events so they do not conflict with user-defined events](https://github.com/andrejewski)
- [ ] [can we remove this weird lookup?](https://github.com/canjs/can-stache/issues/289#issuecomment-340062580)

## Details

The first version of `can-stache` shipped with [CanJS 2.1](https://www.bitovi.com/blog/canjs-2-1-release-notes) nearly 3½ years ago. In that time many new features have been added and countless bugs have been fixed. We've also taught stache to countless meetup attendees and client developers and found many hard edges that make it difficult for new users to learn and difficult for even seasoned CanJS developers to maintain in large apps.

Here are the major changes that we'd like to make in a `can-stache` 5.0 release:

### Remove implicit scope walking

Currently in `can-stache`, any variables you use within magic tags or bindings are looked up in a hierarchy of [scopes](https://canjs.com/doc/can-view-scope.html).

This means that if you render a stache template like

```html
{{#each children}}
    <li>{{name}} is {{age}} years old</li>
{{/each}}
```

with some data like

```js
{
    name: 'Justin',
    age: 33,
    children: [{
        name: 'Ramiya',
        age: 2
    }, {
        name: 'Joffrey'
    }]
}
```

You will get an output that looks like
```html
    <li>Ramiya is 2 years old</li>
    <li>Joffrey is 33 years old</li>
```

It might be obvious what is happening in this simple example, but when this happens in larger apps it is often very difficult to identify.

There are many ways to prevent this from happening; You can use `./` and `../` to explicitly specify the scope you want to read from.

```html
{{#each children}}
    <li>{{./name}} is {{./age}} years old</li>
{{/each}}
```

We are also adding the ability to use hash expressions to explicitly name the variables used in `{{#each}}` and `{{#with}}` statements, so you can do something like:

```html
{{#each children child=value}}
    <li>{{child.name}} is {{child.age}} years old</li>
{{/each}}
```

These things make it possible to do the right thing, so we should remove the current default thing which is almost always not what you want.

If you want to explicitly walk up the scope, you can still do that with `../`.

### Remove automatic function calling

The default behavior in Stache lookups is to execute functions to get their return values.

This confuses people because they expect to have to call a function in order to get its return value. For example, people get confused that both of the following will call `myFunction` and set the `value` of the input to whatever `myFunction` returns.

```
<input value:from="myFunction" type="text">
<input value:from="myFunction()" type="text">
```

This also means that in order to actually pass the function through bindings, you need to use [`@`](https://canjs.com/doc/can-stache/keys/at.html) which people sometimes also find confusing.

In 5.0, we should remove this behavior so that functions are not automatically called and you have to call them yourself in order to get the return value.

### Remove [`%special`](https://canjs.com/doc/can-stache/keys/special.html) keywords

This is detailed in depth in [this issue](https://github.com/canjs/can-stache/issues/304).

### Remove `@`

Since we will no longer call functions by default, we do not need `@` to prevent that behavior. Also discussed in [this issue](https://github.com/canjs/can-stache/issues/47).

### Use `can-parse` for parsing Expressions

This is [#290](https://github.com/canjs/can-stache/issues/290). This isn't a breaking change, but would rewrite all of the code that is used for parsing expressions in can-stache so could be done along with 5.0 in order to reduce the risk of regressions in a patch/minor release.

The purpose of this is to increase the consistency of the behavior of stache expressions by using a structured parser instead of the [while loop and if/else statements](https://github.com/canjs/can-stache/blob/master/src/expression.js#L763-L922) that are used today.

This will reduce the number of issues where certain Expressions do not work in every scenario someone expects and would allow us to more clearly specify when you can use each type of Expression.

For example, here are some issues related to Bracket Expressions not working in different places:

- [Bracket syntax doesn't work as a helper argument](https://github.com/canjs/can-stache/issues/83)
- [Using bracket in a call expression doesn't work](https://github.com/canjs/can-stache/issues/84)
- [allowing you to pass Bracket expression as an argument to a Call expression](https://github.com/canjs/can-stache/pull/86)
- [Bracket expression in an attribute throws an error](https://github.com/canjs/can-stache/pull/79)
- [Adding support for passing bracket expressions as arguments to helpers](https://github.com/canjs/can-stache/pull/89)
- [Making bracket expression work in multi-arg helper](https://github.com/canjs/can-stache/pull/103)
- [Bracket expressions should support another Bracket or Lookup](https://github.com/canjs/can-stache/issues/109)
- [Support a Bracket expression followed by a Lookup
](https://github.com/canjs/can-stache/issues/122)
- [Fix for Call expressions inside of Bracket expressions](https://github.com/canjs/can-stache/pull/146)

These were all related to Bracket Expressions simply because this was the Expression that was added most recently. Similar problems have existed with other types of Expressions that people wanted to use in places that weren't originally anticipated.

Using a structured parser with an actual grammar will make it much easier to avoid issues like this and will allow us to add new features much more quickly.

**Comment by matthewp** [comment]
If special keywords are removed how would you reference, for example, the index within an array?

**Comment by phillipskevin** [comment]
One idea is to use a reserved `scope` keyword. https://github.com/canjs/can-stache/issues/287 describes that.

I've also though about making our helpers look more like arrow functions (instead of using `as`), so you would do something like:

```html
{{#each(todos => (todo, index))}}
    <li>todo {{index}}: {{todo.name}}</li>
{{/each}}
```

**Comment by andrejewski** [comment]
Ember does [`as |value index|`](https://stackoverflow.com/a/19795719) to expose the index of an array. Example:

```html
{{#each model as |item index|}}
  <li>
    Index: {{index}} Content: {{item}}
  </li>
{{/each}}
```

**Comment by matthewp** [comment]
Cool, we have an `as` as well: https://canjs.com/doc/can-stache.helpers.each.html#___eachEXPRESSIONasKEY__FN__else__INVERSE___each__

Maybe this could be made to support some type of destructing syntax...

**Comment by justinbmeyer** [comment]
Do we want to enable people to call functions like:

```
{{myFunc}} vs {{myFunc}}

{{#each thing}}
```

`{{foo}}`

1. check if on current context
2.check helpers
3. walk up scope 

### Instead

1. check on current scope
2. check global helpers? (can we remove options scope?)
3. break

**Comment by justinbmeyer** [comment]
zed: "string"


<component foo:from="zed"/>


can-stache-key 

can.hasOwnProperty(obj,"zed") //-> true

don't warn

**Comment by justinbmeyer** [comment]
`{{getThingFrom id}}`

**Comment by justinbmeyer** [comment]
`{{#each( items, item=boo) }}`

**Comment by phillipskevin** [comment]
can we remove this weird lookup? https://github.com/canjs/can-stache/blob/8a4646db8ab42da2bd1dee3299ad47ffdefa7ea4/expressions/helper.js#L55-L58

**Comment by frank-dspeed** [comment]
@phillipskevin i think this issue needs to get reduced to 5.0 related parts as can stache is already 4 ?
